cmd/dp-observation-importer/main.go

```diff
@@ -82,69 +129,122 @@ func main() {
+		}
 
-	err = kafkaResultProducer.Close(ctx)
-	logError(err)
+		// If observation imported error kafka producer exists, close it
+		if serviceList.ObservationsImportedErrProducer {
+			log.Event(shutdownContext, "closing observation imported error kafka producer", log.INFO, log.Data{"producer": "DimensionExtracted"})
+			observationsImportedErrProducer.Close(shutdownContext)
+			log.Event(shutdownContext, "closed observation imported error kafka producer", log.INFO, log.Data{"producer": "DimensionExtracted"})
+		}
 
-	err = graphDB.Close(ctx)
-	logError(err)
+		// Attempting to close event consumer
+		log.Event(shutdownContext, "closing event wrapper", log.INFO)
+		eventConsumer.Close(shutdownContext)
+		log.Event(shutdownContext, "closing event wrapper", log.INFO)
+
+		// If kafka consumer exists, close it.
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "closing kafka consumer", log.INFO, log.Data{"consumer": "SyncConsumerGroup"})
+			syncConsumerGroup.Close(shutdownContext)
+			log.Event(shutdownContext, "closed kafka consumer", log.INFO, log.Data{"consumer": "SyncConsumerGroup"})
+		}
 
-	err = httpServer.Shutdown(ctx)
-	logError(err)
+		if serviceList.Graph {
+			err = graphDB.Close(ctx)
+			logIfError(ctx, "failed to close graph db", err)
+		}
+	}()
 
-	// cancel the timer in the shutdown context.
-	cancel()
+	// wait for shutdown success (via cancel) or failure (timeout)
+	<-shutdownContext.Done()
 
-	log.Debug("graceful shutdown was successful", nil)
+	log.Event(shutdownContext, "graceful shutdown was successful", log.INFO)
 	os.Exit(0)
 }
 
-func checkForError(err error) {
+// registerCheckers adds the checkers for the provided clients to the healthcheck object
+func registerCheckers(ctx context.Context, hc *healthcheck.HealthCheck,
+	kafkaConsumer *kafka.ConsumerGroup,
+	observationsImportedProducer *kafka.Producer,
+	observationsImportedErrProducer *kafka.Producer,
+	datasetClient dataset.Client,
+	graphDB *graph.DB) (err error) {
+
+	if err = hc.AddCheck("Kafka Consumer", kafkaConsumer.Checker); err != nil {
+		log.Event(ctx, "error adding check for kafka consumer", log.ERROR, log.Error(err))
+	}
+
+	if err = hc.AddCheck("Kafka Import Producer", observationsImportedProducer.Checker); err != nil {
+		log.Event(ctx, "error adding check for kafka import producer", log.ERROR, log.Error(err))
+	}
+
+	if err = hc.AddCheck("Kafka Error Producer", observationsImportedErrProducer.Checker); err != nil {
+		log.Event(ctx, "error adding check for kafka error producer", log.ERROR, log.Error(err))
+	}
+
+	if err = hc.AddCheck("dataset API", datasetClient.Checker); err != nil {
+		log.Event(ctx, "error adding check dataset client", log.ERROR, log.Error(err))
+	}
+
+	if err = hc.AddCheck("graph db", graphDB.Driver.Checker); err != nil {
+		log.Event(ctx, "error adding check dataset client", log.ERROR, log.Error(err))
+	}
+
+	return
```

Comment:
I think you may need a return statement in each `if` block here. Currently this has _fall through_ logic for example if `Kafka Consumer` return an error it would assign it to `err` and then continue to `Kafka Import Producer`. If this also returned an error it will overwrite the first error and we will lose visibility of the first error. 

I think we want it to fail fast here.

Reply:
The purpose of it written like this is so there is an error log for each failed checker created for health check without returning early. This way we know which checkers failed instead of starting the app several times to find each checker in turn has failed.

Where registerCheckers() function is called, it will just exit the application when it receives an error (there isn't a error log as these would have been created in the actual function)

Reply:
Sorry you are correct, total oversight of mine; we don't want to return straight away but should track errors and return if there is one. David has since updated this